sol/main.c

```diff
@@ -0,0 +1,459 @@
+
+	if (pkt_len > req_queue->tb_credits)
+		return false;
+
+	req_queue->tb_credits -= pkt_len;
+	return true;
+}
+
+static void
+dequeue_reqs(struct sol_config *sol_conf, uint8_t tx_port,
+	uint16_t tx_queue)
+{
+	struct req_queue *req_queue = &sol_conf->req_queue;
+	struct priority_req *entry, *next;
+
+	struct rte_mbuf *pkts_out[sol_conf->deq_burst_size];
+	uint32_t nb_pkts_out = 0;
+	uint16_t total_sent = 0;
+
+	/* Get an up-to-date view of our credits. */
+	time_resync(req_queue);
+	credits_update(req_queue);
+
+	list_for_each_entry_safe(entry, next, &req_queue->head, list) {
+		struct rte_mbuf *pkt = entry->pkt;
+
+		if (!credits_check(req_queue, pkt)) {
+			/*
+			 * XXX When under an attack, we may not want to
+			 * log this because it could become expensive.
+			 */
+			RTE_LOG(NOTICE, GATEKEEPER,
+				"sol: out of request bandwidth\n");
+			goto out;
+		}
+
+		if (req_queue->len == 1 || (entry->priority != next->priority))
+			req_queue->priorities[entry->priority] = NULL;
+		list_del(&entry->list);
+		mb_free_entry(&sol_conf->mb, entry);
+		req_queue->len--;
+
+		pkts_out[nb_pkts_out++] = pkt;
+
+		/* Advance port time. */
+		req_queue->time += pkt->pkt_len +
+			RTE_SCHED_FRAME_OVERHEAD_DEFAULT;
+	}
+
+out:
+	if (list_empty(&req_queue->head))
+		req_queue->lowest_priority = req_queue->num_priorities;
+	else {
+		struct priority_req *first = list_first_entry(&req_queue->head,
+			struct priority_req, list);
+		req_queue->highest_priority = first->priority;
+	}
+
+	/* We cannot drop the packets, so re-send. */
+	do {
+		uint16_t sent = rte_eth_tx_burst(tx_port, tx_queue,
+			pkts_out + total_sent, nb_pkts_out);
+		total_sent += sent;
+		nb_pkts_out -= sent;
+	} while (nb_pkts_out > 0);
+}
+
+static int
+req_queue_init(struct req_queue *req_queue)
+{
+	struct rte_eth_link link;
+	uint32_t cycles_per_byte;
+	uint64_t rate;
+	int ret;
+
+	req_queue->len = 0;
+	req_queue->num_priorities = GK_NUM_REQ_PRIORITIES;
+	req_queue->highest_priority = 0;
+	req_queue->lowest_priority = req_queue->num_priorities;
+
+	/* Token bucket. */
+	ret = rte_approx(REQ_QUEUE_BW_RATE, GATEKEEPER_TB_RATE_CONFIG_ERR,
+		&req_queue->tb_credits_per_period, &req_queue->tb_period);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER, "sol: could not approximate the request queue's allocated bandwidth\n");
+		return ret;
+	}
+
+	/* Calculate link speed in bytes per second. */
+	rte_eth_link_get(sol_conf->net->back.id, &link);
+	rate = (uint64_t)link.link_speed * 1000 * 1000 / 8;
+	cycles_per_byte = (rte_get_tsc_hz() << GATEKEEPER_TIME_SHIFT) / rate;
+	req_queue->inv_cycles_per_byte = rte_reciprocal_value(cycles_per_byte);
+
+	/* The DPDK sched library starts the bucket half full. */
+	req_queue->tb_size = rate * REQ_QUEUE_BW_RATE;
+	req_queue->tb_credits = req_queue->tb_size / 2;
+	req_queue->tb_time = 0;
+
+	req_queue->time_cpu_cycles = rte_get_tsc_cycles();
+	req_queue->time_cpu_bytes = 0;
+	req_queue->time = 0;
+
+	INIT_LIST_HEAD(&req_queue->head);
```

Comment:
Technically no, since we are allocating the entire configuration structure (which includes `req_queue`) using `rte_calloc()`, which zeroes the memory. Still, I've explicitly set some fields to zero in this initialization function, so I will zero the `priorities` array to make it consistent.